src/rgw/rgw_env.cc

```diff
@@ -54,7 +54,12 @@ const char *rgw_conf_get(const map<string, string, ltstr_nocase>& conf_map, cons
 
 const char *RGWEnv::get(const char *name, const char *def_val) const
 {
-  return rgw_conf_get(env_map, name, def_val);
+  const char *result = rgw_conf_get(env_map, name, def_val);
+  if (result == nullptr) {
```

Comment:
Wouldn't it be cleaner to change the default value of def_val to an empty string instead? That way if a caller passes nullptr for def_val, this method will not mess with the expected return value.

Reply:
Thanks @vedanshbhartia for your feedback, I've just made another commit fixes this.

Reply:
i looked at the calling code: rgw_rest.h, rgw_auth_s3.h/cc, rgw_rest_client.cc
all of these places interpret `def_val=nullptr` as "not found without default value".
this means, that any change to the return value would break the existing logic.

so, i think there could be several options:
* leave the things as they are (maybe add a commend at rgw_common.h saying that this function returns nullptr if value is not found and default value is not provided. and still have the risk of dereferencing a nullptr
* change the return type to `cosnt std::string&`. since the internal map already holds `std::string`, it may make sense to return it from the function. and pass empty string as the default value
* or, to avoid confusion between an empty string, and a missing value, i would not go for an empty string as the missing value, instead, i would return `std::optiona<std::string>` which clearly indicated whether a value exists or not, and have two overloads of the function, one with a default, and one without

Reply:
Thanks @yuvalif for looking into the calling code and providing options for the return type of the function. I appreciate your thorough analysis.

Regarding the options you suggested, I understand the potential confusion between an empty string and a missing value. I think using `std::optional<std::string>` as the return type is a good solution as it clearly indicates whether a value exists or not. Having two overloads of the function, one with a default and one without, would also be helpful.

I've chosen the second option and made a commit for that.

Please let me know if you need any further changes or if everything looks good now.

Reply:
there's a [`RGWHTTPArgs::get_optional()`](https://github.com/ceph/ceph/blob/9ec470a/src/rgw/rgw_common.h#L360-L361) that behaves this way, except instead of `std::optional<std::string>` it returns `boost::optional<const std::string&>` to return a reference instead of a copy